Lib/imaplib.py

```diff
@@ -1279,6 +1389,251 @@ def print_log(self):
+    subject to change.  Calling code can rely on its context management,
+    iteration, and public method to remain stable, but should not
+    subclass, instantiate, or otherwise directly reference the class.
+    """
+
+    def __init__(self, imap, duration=None):
+        if 'IDLE' not in imap.capabilities:
+            raise imap.error("Server does not support IMAP4 IDLE")
+        self._duration = duration
+        self._imap = imap
+        self._tag = None
+        self._sock_timeout = None
+        self._old_state = None
+
+    def __enter__(self):
+        imap = self._imap
+        assert not (imap._idle_responses or imap._idle_capture)
+
+        if __debug__ and imap.debug >= 4:
+            imap._mesg(f'idle start duration={self._duration}')
+
+        try:
+            # Start capturing untagged responses before sending IDLE,
+            # so we can deliver via iteration any that arrive while
+            # the IDLE command continuation request is still pending.
+            imap._idle_capture = True
+
+            self._tag = imap._command('IDLE')
+            # As with any command, the server is allowed to send us unrelated,
+            # untagged responses before acting on IDLE.  These lines will be
+            # returned by _get_response().  When the server is ready, it will
+            # send an IDLE continuation request, indicated by _get_response()
+            # returning None.  We therefore process responses in a loop until
+            # this occurs.
+            while resp := imap._get_response():
+                if imap.tagged_commands[self._tag]:
+                    raise imap.abort(f'unexpected status response: {resp}')
+
+            if __debug__ and imap.debug >= 4:
+                prompt = imap.continuation_response
+                imap._mesg(f'idle continuation prompt: {prompt}')
+        except BaseException:
+            imap._idle_capture = False
+            raise
+
+        self._sock_timeout = imap.sock.gettimeout() if imap.sock else None
+        if self._sock_timeout is not None:
+            imap.sock.settimeout(None)  # Socket timeout would break IDLE
+
+        self._old_state = imap.state
+        imap.state = 'IDLING'
+
+        return self
+
+    def __iter__(self):
+        return self
+
+    def _wait(self, timeout=None):
+        # Block until the next read operation should be attempted, either
+        # because data becomes availalable within 'timeout' seconds or
+        # because the OS cannot determine whether data is available.
+        # Return True when a blocking read() is worth trying
+        # Return False if the timeout expires while waiting
+
+        imap = self._imap
+        if timeout is None:
+            return True
+        if imap._readbuf:
+            return True
+        if timeout <= 0:
+            return False
+
+        if imap.sock:
+            fileobj = imap.sock
+        elif platform.system() == 'Windows':
```

Comment:
Can't we use `os.name == 'nt'` instead? that's how we generally check Windows-related code.